### ISSUE TITLE
[BUG] Tests crash out on ctrl-c

### ISSUE THREAD

**Issue by rotu** [npm/cli#6766]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

Interrupting tests with ctrl-c results in an error:

> TypeError: The "code" argument must be of type number. Received type string ('128SIGINT')
>     at process.set [as exitCode] (node:internal/bootstrap/node:123:9)
>     at ChildProcess.<anonymous> (/Users/dan/Source/npm-cli/node_modules/nyc/node_modules/foreground-child/index.js:63:22)
>     at ChildProcess.emit (node:events:514:28)
>     at ChildProcess.emit (node:domain:489:12)
>     at maybeClose (node:internal/child_process:1105:16)
>     at Process.ChildProcess._handle.onexit (node:internal/child_process:305:5)

This seems to come from bad logic [adding a number to a string](https://github.com/tapjs/foreground-child/blob/efba8bb6d7607b5c3b0d3e3897c73727577c8b45/index.js#L61-L63) in `foreground-child@2.0.0` (outdated) via `nyc@15.1.0` (latest, but unmaintained).

### Expected Behavior

Tests should exit cleanly.

### Steps To Reproduce

1. In Mac or Windows
2. Run `npm test`
3. Use ctrl-c to abort the tests
4. Observe error

### Environment

- npm: 10.0.0
- Node.js: 20.6.0
- OS Name: macOS
- System Model Name: Macbook Air
- npm config:
```ini
; "builtin" config from /opt/homebrew/lib/node_modules/npm/npmrc

prefix = "/opt/homebrew" 

; "user" config from /Users/dan/.npmrc

auto-install-peers = true 

; "project" config from /Users/dan/Source/npm-cli/.npmrc

package-lock = true 

; node bin location = /opt/homebrew/Cellar/node/20.6.0/bin/node
; node version = v20.6.0
; npm local prefix = /Users/dan/Source/npm-cli
; npm version = 9.8.1
; cwd = /Users/dan/Source/npm-cli
; HOME = /Users/dan
; Run `npm config ls -l` to show all defaults.
```

**Comment by rotu** [comment]
FYI, @lukekarrys. I wonder if whatever necessitated https://github.com/npm/cli/commit/b8a5764db60cb2b3edb7887cf4379f48411a0b55 is related to this issue.

**Comment by lukekarrys** [comment]
I haven't looked into this too deeply yet, but I believe this is due to a change in Node 20 of new restrictions on `process.code`. I think this should only affect our tests for now, but still something that should be fixed once the current batch of priority bugs in npm 10 are squashed.

**Comment by rotu** [comment]
Ah. Can confirm this happens on my machine in Node 20 but not Node 18.

Node 19 gives a deprecation warning:

<pre>
(node:38601) [DEP0164] DeprecationWarning: Implicit coercion to integer for exit code is deprecated.
    at process.set [as exitCode] (node:internal/bootstrap/node:118:9)
    at ChildProcess.<anonymous> (/Users/dan/Source/npm-cli/node_modules/nyc/node_modules/foreground-child/index.js:63:22)
    at ChildProcess.emit (node:events:513:28)
    at ChildProcess.emit (node:domain:489:12)
    at maybeClose (node:internal/child_process:1098:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:304:5)
</pre>

**Comment by rotu** [comment]
@lukekarrys Actually, you may want to bump *up* the priority. Further experimentation shows that, in Node 18, the process actually exits with exit code 0. This could very well be causing tests that should fail to accidentally pass!

```shell
node -e 'process.exitCode=42; process.exit()'; echo $?
# 42
node -e 'process.exitCode=42+"SIGINT"; process.exit()'; echo $?
# 0
```

**Comment by rotu** [comment]
FYI @isaacs, this seems important to your work on tapjs.

**Comment by isaacs** [comment]
I already have much more sophisticated signal and error handling code in tap 18. Unfortunately, I'm a bit delayed in shipping it, because the changes to the loader system in node have required quite a lot of work to get everything converted over. The mockImport method is almost done being converted, and I'm waiting on https://github.com/TypeStrong/ts-node/pulls/2009 to land for ts-node to work there.

Once those are done, I can ship tap 18, and ignore this and all NYC errors forever 😅 

Can you get it to happen in foreground-child v3? If not, I'm inclined to say it's just `WONTFIX`, not out of any malice, just because the 2.x codebase is long out of date.

**Comment by isaacs** [comment]
Omg, I just looked at the link you posted. Adding 128 to the signal byte to get it to exit properly. Those were the ""good"" old days, for sure.

**Comment by rotu** [comment]
> I already have much more sophisticated signal and error handling code in tap 18. Unfortunately, I'm a bit delayed in shipping it, because the changes to the loader system in node have required quite a lot of work to get everything converted over. The mockImport method is almost done being converted, and I'm waiting on https://github.com/TypeStrong/ts-node/pull/2009 to land for ts-node to work there.
> 
> Once those are done, I can ship tap 18, and ignore this and all NYC errors forever 😅
> 
> Can you get it to happen in foreground-child v3? If not, I'm inclined to say it's just `WONTFIX`, not out of any malice, just because the 2.x codebase is long out of date.

Awesome! I didn't know you were ditching `nyc` - that makes this much less a problem moving forward.

It's also good that this doesn't fail so quietly on Node 20. Why it *ever* failed silently makes me fear for humanity a little bit.

> Omg, I just looked at the link you posted. Adding 128 to the signal byte to get it to exit properly. Those were the ""good"" old days, for sure.

Heh. It seems like the `close` handler *always* [got a string for the signal](https://nodejs.org/docs/latest-v0.10.x/api/child_process.html#child_process_event_close), so setting the exit code based on a signal never worked quite [as intended](https://nodejs.org/docs/latest-v0.10.x/api/process.html#process_signal_events).

**Comment by isaacs** [comment]
> Heh. It seems like the close handler always [got a string for the signal](https://nodejs.org/docs/latest-v0.10.x/api/child_process.html#child_process_event_close), so setting the exit code based on a signal never worked quite [as intended](https://nodejs.org/docs/latest-v0.10.x/api/process.html#process_signal_events).

Kids these days thinking node 0.10 was the first node version, get off my lawn etc. Yeah, istr that is copy-pasta from an `on('exit')` handler code dating back to around node 0.1.100-ish days. It's never worked as written, and frankly surprising to me that it was never caught until now. In any case, exiting with a status code byte with the high bit set to indicate a signal exit is an ancient idiom, and has never worked on Windows.

> It's also good that this doesn't fail so quietly on Node 20. Why it ever failed silently makes me fear for humanity a little bit.

You and me both.
